### PR TITLE
WIP: Optionally build Crypto and SslStream using OpenSSL on Windows (part1)

### PR DESCRIPTION
**Pull request by wfurt**
This is not meant as product change. However when built locally, it allows to use VS analyzers and profiling tools as well as any other Windows tools on "Linux" code path. (including native parts)  

ssl-open

This if first part covering PAL and low-level crypto. I have working PoC but managed code still needs some more work and I wanted early feedback. 

This change has:
- builds native System.Security.Cryptography.Nartive parts from Unix pal_* files.
- add missing POSIX opendir/readdir/closedir on top of Windows API

-adds ErrorInfo for Windows.
This part I'm not quite sure as the note mentioned platform independent API but it currently exists only for Unix. For now, it assumes 1:1 mapping to Windows codes. Crypto interop really only needs Error.SUCCESS and rest is only used for logging.

There may be better way how to enable this but for now I use existence of OPENSSLDIR pointing to local OpenSSL install. This will also allow to build agains different versions and iot does not require OpenSSL to be installed to system location. 

Second part I'm not sure about is BCL0015. It did complain about pinvokes even if we do them on other platforms. I don't know if this is only because we only use the analyzer on Windows or if there is any other reason. So I suppress them for now since this is optional build feature and it would not be used in production build.

### CODE REVIEW

**Comment by bartonjs** [comment]
I'm not sure there's anything that can move me from no to yes on this.

File: src/libraries/Native/Windows/Common/pal_safecrt.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (jkotas) [inline src/libraries/Native/Windows/Common/pal_safecrt.h]: All this duplication is unfortunate. 

You may want to break this change into smaller steps that are more acceptable. 

We have more and more of the Unix libraries coming to Windows. Compression was on Windows in forever, Globalization is over (cc @safern) and now you would like to have build crypto (as optional non-shipping test-only). So the first step may be to get rid of the Windows/Unix separation under libraries/Native and more everything to the same level.

**Comment by wfurt** [comment]
thanks for feedback. I'll keep it open to carry conversation and I'll explore some of the suggestions.

**Comment by stephentoub** [comment]
@wfurt, closing per our discussion offline.

**Pull request closed without merging**